test/features/email/presentation/controller/single_email_controller_test.dart

```diff
@@ -363,4 +375,137 @@ void main() {
       debugDefaultTargetPlatformOverride = null;
     });
   });
+
+  group('_parseCalendarEventAction method test:', () {
+    final calendarEventDataSource = MockCalendarEventDataSource();
+    final calendarEventRepository = CalendarEventRepositoryImpl(
+      {DataSourceType.network :calendarEventDataSource},
+      HtmlDataSourceImpl(
+        HtmlAnalyzer(HtmlTransform(
+          MockDioClient(),
+          const HtmlEscape(),
+        )),
+        CacheExceptionThrower(),
+      ),
+    );
+
+    setUp(() {
+      Get.put(ParseCalendarEventInteractor(calendarEventRepository));
+    });
+
+    tearDown(() {
+      Get.delete<ParseCalendarEventInteractor>();
+    });
+
+    test(
+      'should transform all calendar event description url to a tag '
+      'and all new line to <br> tag',
+    () async {
+      // arrange
+      const eventDescription = '\nhttps://example1.com\nhttps://example2.com';
+      const expectedEventDescription = '<html><head></head><body>'
+        '<br>'
+        '<a href="https://example1.com" target="_blank" rel="noreferrer">example1.com</a>'
+        '<br>'
+        '<a href="https://example2.com" target="_blank" rel="noreferrer">example2.com</a>'
+        '</body></html>';
+      final blobId = Id('abc123');
+      final calendarEvent = CalendarEvent(
+        description: eventDescription,
+      );
+      final blobCalendarEvents = [
+        BlobCalendarEvent(
+          blobId: blobId,
+          calendarEventList: [calendarEvent],
+        ),
+      ];
+      when(calendarEventDataSource.parse(any, any))
+        .thenAnswer((_) async => blobCalendarEvents);
+
+      when(mailboxDashboardController.selectedEmail).thenReturn(Rxn(PresentationEmail()));
+      when(mailboxDashboardController.emailUIAction).thenReturn(Rxn(EmailUIAction()));
+      when(mailboxDashboardController.viewState).thenReturn(Rx(Right(UIState.idle)));
+      when(mailboxDashboardController.accountId).thenReturn(Rxn(AccountFixtures.aliceAccountId));
+      when(emailSupervisorController.scrollPhysicsPageView).thenReturn(Rxn());
+
+      singleEmailController.onInit();
+      mailboxDashboardController.accountId.refresh();
+      
+      // act
+      singleEmailController.parseCalendarEventAction(
+        accountId: AccountFixtures.aliceAccountId,
+        blobIds: {blobId},
+      );
+      await untilCalled(calendarEventDataSource.parse(any, any));
+      await Future.delayed(Duration.zero);
```

Comment:
Why do we need delayed while duration is zero?

Reply:
If not, the `expect` will not wait for the transformers to do its work before verifying. `await Future.delayed(Duration.zero);` simply add another `Future` to main event queue of Flutter while transformers is executing, so that if this `await Future.delayed(Duration.zero);` is finished, the `transformers` future is sure to finished too.

Reply:
This is too confusing. Let's try to handle it in a more reasonable way.